### PR TITLE
[Feature] add SFSegNet head

### PR DESCRIPTION
**Pull request by MengzhangLI**
Implementation of [Semantic Flow for Fast and Accurate SceneParsing](https://arxiv.org/pdf/2002.10120v2.pdf).
Modified from the [official repository](https://github.com/lxtGH/SFSegNets).

Done: 
- unitest

Todo:

- design the config interface and files
- more experiments especially model with res18 backbone.

### CODE REVIEW

File: mmseg/models/decode_heads/sfnet_head.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+        #
+        # For unitest, otherwise may raise "UnboundLocalError"
+        fpn_inplanes = [256, 512, 1024, 2048]
+        fpn_dim = 256
+        if self.channels == 256:
+            fpn_inplanes = [256, 512, 1024, 2048]
+            fpn_dim = 256
+        elif self.channels == 64:
+            fpn_inplanes = [64, 128, 256, 512]
```
Comment (MengzhangLI) [inline mmseg/models/decode_heads/sfnet_head.py:109]: This `if...else` function made `fpn_inplanes` and `fpn_dim` not a global variable thus if not been specified, error "UnboundLocalError" may raise. Thus I just use resnet-50 channel number for unitest.

File: configs/sfnet/sfnet_r18-d8_512x1024_80k_cityscapes.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+_base_ = './sfnet_r50-d8_512x1024_80k_cityscapes.py'
+model = dict(
+    pretrained='open-mmlab://resnet18_v1c',
```
Comment (Junjun2016) [inline configs/sfnet/sfnet_r18-d8_512x1024_80k_cityscapes.py:3]: Please check whether the r18 is v1c or v1d.

File: mmseg/models/decode_heads/sfnet_head.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            self.channels,
+            3,
+            padding=1,
+            bias=True,
+            conv_cfg=self.conv_cfg,
+            norm_cfg=self.norm_cfg,
+            act_cfg=self.act_cfg)
+
+        #
+        # For unitest, otherwise may raise "UnboundLocalError"
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/sfnet_head.py]: We use unittests to ensure the correctness of the current code and future changes to other code will not affect the current code, instead of changing the current code to pass the unittests.

File: mmseg/models/decode_heads/sfnet_head.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        elif self.channels == 64:
+            fpn_inplanes = [64, 128, 256, 512]
+            fpn_dim = 64
+
+        self.fpn_in = []
+        for fpn_inplane in fpn_inplanes[:-1]:
+            self.fpn_in.append(
+                nn.Sequential(
+                    nn.Conv2d(fpn_inplane, fpn_dim, 1),
+                    nn.BatchNorm2d(fpn_dim), nn.ReLU(inplace=False)))
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/sfnet_head.py]: Merge these to ConvModule.

File: mmseg/models/decode_heads/sfnet_head.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+                nn.Sequential(
+                    nn.Conv2d(
+                        fpn_dim,
+                        fpn_dim,
+                        kernel_size=3,
+                        stride=1,
+                        padding=1,
+                        bias=False),
+                    nn.BatchNorm2d(fpn_dim),
+                    nn.ReLU(inplace=True),
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/sfnet_head.py]: The same.

File: mmseg/models/decode_heads/sfnet_head.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        self.fpn_out = nn.ModuleList(self.fpn_out)
+        self.fpn_out_align = nn.ModuleList(self.fpn_out_align)
+        self.conv_last = nn.Sequential(
+            nn.Conv2d(
+                len(fpn_inplanes) * fpn_dim,
+                fpn_dim,
+                kernel_size=3,
+                stride=1,
+                padding=1,
+                bias=False), nn.BatchNorm2d(fpn_dim), nn.ReLU(inplace=True))
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/sfnet_head.py]: The same.

File: tests/test_models/test_heads/test_sfnet_head.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        channels=256,
+        num_classes=19,
+        pool_scales=(1, 2, 3, 6))
+    if torch.cuda.is_available():
+        head, inputs = to_cuda(head, inputs)
+    assert head.psp_modules[0][0].output_size == 1
+    assert head.psp_modules[1][0].output_size == 2
+    assert head.psp_modules[2][0].output_size == 3
+    outputs = head(inputs)
+    assert outputs.shape == (1, head.num_classes, 45, 45)
```
Comment (Junjun2016) [inline tests/test_models/test_heads/test_sfnet_head.py:44]: We need more unittests to ensure that the intermediate variables are set correctly.

**Comment by Junjun2016** [comment]
Hi @MengzhangLI 
Please resolve the comments.

**Comment by Junjun2016** [comment]
Hi @MengzhangLI 
For every PR, you should checkout to the master first, then pull the upstream master, and finally checkout a new branch.
It can avoid too many necessary committed histories.

File: mmseg/models/decode_heads/sfnet_head.py [hunk 8]
```diff
@@ -143,26 +151,56 @@ def forward(self, x):
...
+        """Implementation of Warp Procedure in Fig 3(b) of original paper,
+        which is between Flow Field and High Resolution Feature Map.
+
+        Args:
+            input (Tensor): High Resolution Feature Map.
+            flow (Tensor): Semantic Flow Field that will give dynamic
+            indication about how to align these two feature maps effectively.
+            size (Tuple): Shape of height and width of output.
+
+        For example, in cityscapes 1025x2048 dataset with ResNet18 config,
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/sfnet_head.py]: [Returns](https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/activation.py#L88)

File: configs/_base_/models/sfnet_r50-d8.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+model = dict(
+    type='EncoderDecoder',
+    pretrained='open-mmlab://resnet50_v1c',
+    backbone=dict(
+        type='ResNetV1d',
+        depth=50,
+        num_stages=4,
+        out_indices=(0, 1, 2, 3),
+        dilations=(1, 1, 2, 4),
+        strides=(1, 2, 2, 2),
```
[the remark this answers is not in the document]
Comment (Junjun2016) [reply configs/_base_/models/sfnet_r50-d8.py:12]: Rename config from d8 to d32.

**Comment by MengzhangLI** [comment]
Re-implementing SFNet ResNet50 of PaddleSeg.

Inference metric is done.
image

Note: 
`stride` in model of paddleseg sfnet Res50 is `(1, 2, 1, 1)`, where `original paper` is `(1, 2, 2, 2)`
That's why the performance of its [sfnet res50](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.2/configs/sfnet) `81.49` is better than [paper](https://github.com/lxtGH/SFSegNets) `80.74`.

**Comment by MengzhangLI** [comment]
Re-implementing SFNet ResNet18 of [PaddleSeg](https://github.com/PaddlePaddle/PaddleSeg/tree/dd6e03c138f7183ed4b005bf1979ea3b9cc8fc13/configs/sfnet).

Note: there are some potential risks for paddleseg users:
(1) Actual Output Stride(OS) of Paddleseg SFNet ResNet18 model is not **8** but **32**. Although **32** is setting of original paper,  It is unintentionally correct settings because paddleseg setting is **8** from [here](https://github.com/PaddlePaddle/PaddleSeg/blob/405f49eb6ee74cc82717e501bdb93dc5306ef689/configs/sfnet/sfnet_resnet18_os8_cityscapes_1024x1024_80k.yml#L10)

The reason of **OS = 32** is when building backbone models, paddleseg would skip dilation rate from [here](https://github.com/PaddlePaddle/PaddleSeg/blob/dd6e03c138f7183ed4b005bf1979ea3b9cc8fc13/paddleseg/models/backbones/resnet_vd.py#L308) when `layers` <= 50.

Thus, stride of SFNet of Paddleseg are `(1, 2, 1, 1), os = 8` (Res50) and  `(1, 2, 2, 2), os = 32` (Res18), respectively.

(2) When using ResNetv1d with 18 layers, paddleseg would add a `short` downsample operations (conv + BN) at first layer, where official usage ([mmsegmentation](https://github.com/open-mmlab/mmsegmentation/blob/ee66167999212593451f9cbf0746c2b5b79cc0ab/mmseg/models/utils/res_layer.py#L43) adopted) would not add. Thus it would have five more variables:
`backbone.bb_0_0.short._conv.weight (64, 64, 1, 1)`
`backbone.bb_0_0.short._batch_norm.weight (64,)`
`backbone.bb_0_0.short._batch_norm.bias (64,)`
`backbone.bb_0_0.short._batch_norm._mean (64,)`
`backbone.bb_0_0.short._batch_norm._variance (64,)`

Anyway, inference metric is DONE.

image

**Comment by MengzhangLI** [comment]
Related SFNet model: https://github.com/azxj/SFNet

**Comment by jyang68sh** [comment]
Hi I just runned sfnet_r18-d32_512x1024_80k_cityscapes.py and only get mIoU of 73.92. I am wondering the cause. Did you change the config?
